docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,505 @@
+        - [Modification to existing API Types](#modification-to-existing-api-types)
+            - [Cluster](#cluster)
+        - [Validations](#validations)
+            - [ClusterClass](#clusterclass-2)
+            - [Cluster](#cluster-1)
+        - [Behaviors](#behaviors)
+            - [Create a new Cluster using ClusterClass object](#create-a-new-cluster-using-clusterclass-object)
+            - [Update an existing Cluster using ClusterClass](#update-an-existing-cluster-using-clusterclass)
+        - [Provider implementation](#provider-implementation)
+            - [For infrastructure providers](#for-infrastructure-providers)
+            - [For Control plane providers](#for-control-plane-providers)
+      - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+      - [Test Plan [optional]](#test-plan-optional)
+      - [Graduation Criteria [optional]](#graduation-criteria-optional)
+      - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+### ClusterClass
+A collection of templates that define a topology (control plane and machine deployments) to be used in one or more Managed Clusters.
+
+### Managed Cluster topology
+A Managed Cluster is a Cluster that provides a single control point to manage it's own topology; the managed cluster topology is defined by a ClusterClass.
+
+### Worker class
+A worker class is a collection of templates that define a set of worker nodes in the cluster.
+
+
+## Summary
+
+This proposal introduces a new ClusterClass object which will be used to provide easy stamping of clusters of similar shapes. It serves as a collection of template resources which are used to generate one or more clusters of the same flavor.
+
+We're enhancing the Cluster CRD and controller to use a ClusterClass resource to provision the underlying objects that compose a cluster. Additionally, the Cluster provides a single control point to manage the Kubernetes version, worker pools, labels, replicas, and so on.
+
+## Motivation
+
+Currently, Cluster API does not expose a native way to provision multiple clusters of the same configuration. The ClusterClass object is supposed to act as a collection of template references which can be used to create managed topologies.
+
+Today, the Cluster object is a logical grouping of components which describe an underlying cluster. The user experience to create a cluster requires the user to create a bunch of underlying resources such as KCP (control plane provider), MachineDeployments, and infrastructure or bootstrap templates for those resources which logically end up representing the cluster. Since the cluster configuration is spread around multiple components, upgrading the cluster version is hard as it requires changes to different fields in different resources to perform an upgrade.
```

Comment:
Can we elaborate on why is a clusterClass necessary or the solution choosen to have a single point to manage version?